File: BOJ/13460_2.py
# 구슬 탈출 2

import time

prev_time = time.time_ns()



import sys
# sys.stdin = open('input.txt', 'r')


read = sys.stdin.readline

ri, rj = -1, -1
bi, bj = -1, -1
hi, hj = -1, -1
mp = []
visited = []


def printMap(ball_pos):
    ri, rj, bi, bj = ball_pos
    for i in range(len(mp)):
        for j in range(len(mp[0])):
            if i == ri and j == rj:
                print('R', end=' ')
            elif i == bi and j == bj:
                print('B', end=' ')
            else:
                print(mp[i][j], end=' ')
        print()


# dir : 북, 동, 남, 서
def tilting(dir, ball_pos):
    result = False
    ri, rj, bi, bj = ball_pos
    di = [-1, 0, 1, 0]
    dj = [0, 1, 0, -1]

    new_bi = bi
    new_bj = bj
    bcnt = 0
    while mp[new_bi][new_bj] != '#':
        if mp[new_bi][new_bj] == 'O':
            return -1
        new_bi += di[dir]
        new_bj += dj[dir]
        bcnt += 1
    new_bi -= di[dir]
    new_bj -= dj[dir]

    new_ri = ri
    new_rj = rj
    rcnt = 0
    while mp[new_ri][new_rj] != '#':
        if mp[new_ri][new_rj] == 'O':
            return 1
        new_ri += di[dir]
        new_rj += dj[dir]
        rcnt += 1
    new_ri -= di[dir]
    new_rj -= dj[dir]

    if new_ri == new_bi and new_rj == new_bj:
        if bcnt > rcnt:
            new_bi -= di[dir]
            new_bj -= dj[dir]
        else:
            new_ri -= di[dir]
            new_rj -= dj[dir]

    ball_pos[:] = [new_ri, new_rj, new_bi, new_bj]
    return 0


def DFS(dir, ball_pos, cnt):
    res = 11
    # print('ball_pos', ball_pos, 'cnt', cnt)
    # printMap(ball_pos)

    for i in range(4):
        new_ball_pos = ball_pos[:]
        tilting_res = tilting(i, new_ball_pos)
        if tilting_res == -1:
            continue
        elif tilting_res == 1:
            return cnt + 1

        new_ri, new_rj, new_bi, new_bj = new_ball_pos
        if visited[new_ri][new_rj][new_bi][new_bj] <= cnt + 1:
            continue
        visited[new_ri][new_rj][new_bi][new_bj] = cnt + 1

        DFS_res = DFS(i, new_ball_pos, cnt + 1)
        if DFS_res != -1:
            # print('goal!', DFS_res)
            res = min(res, DFS_res)
    return res


def solution(N, M):
    ball_pos = [ri, rj, bi, bj]
    res = DFS(0, ball_pos, 0)
    if res == 11:
        return -1
    return res


N, M = map(int, read().split())
for i in range(N):
    li = list(read().strip())
    for j in range(M):
        if li[j] == 'R':
            ri, rj = i, j
            li[j] = '.'
        elif li[j] == 'B':
            bi, bj = i, j
            li[j] = '.'
        elif li[j] == 'O':
            hi, hj = i, j
    mp.append(li)
visited = [[[[11 for _ in range(M)] for _ in range(N)] for _ in range(M)] for _ in range(N)]
print(solution(N, M))



print('time', time.time_ns() - prev_time)

File: BOJ/13460_2.cpp
#include <iostream>
#include <cstdio>

using namespace std;

int N,M,red_ball[2],blue_ball[2];
char map_arr[10][10];
int mem_arr[10][10][10][10] = {0};
int ans = 11;

int di[4]={-1,1,0,0};
int dj[4]={0,0,-1,1};


void printMap(int ri,int rj,int bi,int bj){
    for(int i=0;i<N;i++){
        for(int j=0;j<M;j++){
            if (i==ri && j==rj){
                cout << "R" << "\t";
            } else if (i==bi && j==bj){
                cout << "B" << "\t";
            } else {
                cout << map_arr[i][j] << "\t";
            }
        }
        cout << endl;
    }
    cout << endl;
}


int move(int ri,int rj,int bi,int bj,int cnt){
    printMap(ri,rj,bi,bj);

    if (cnt == 11){
        return -1;
    }

    if (cnt >= ans){
        return -1;
    }

    
    for(int k=0;k<4;k++){
        int mr=0,mb=0;
        int new_ri=ri, new_rj=rj, new_bi=bi, new_bj=bj;

        while(map_arr[new_bi+di[k]][new_bj+dj[k]] != '#'){
            if (map_arr[new_bi+di[k]][new_bj+dj[k]] == 'O'){
                return -1; // 실패
            }
            
            new_bi+=di[k], new_bj+=dj[k];
            mb++;
        }

        while(map_arr[new_ri+di[k]][new_rj+dj[k]] != '#'){
            if (map_arr[new_ri+di[k]][new_rj+dj[k]] == 'O'){
                ans = ans > cnt ? cnt : ans;
                return 1;
            }
            
            new_ri+=di[k], new_rj+=dj[k];
            mr++;
        }

        if(new_ri==new_bi && new_rj==new_bj){
            if(mr > mb){
                new_ri -= di[k];
                new_rj -= dj[k];
            } else {
                new_bi -= di[k];
                new_bj -= dj[k];
            }
        }

        if(mem_arr[new_ri][new_rj][new_bi][new_bj] == 0 || 
                mem_arr[new_ri][new_rj][new_bi][new_bj] > cnt){
            mem_arr[new_ri][new_rj][new_bi][new_bj] = cnt;
            move(new_ri,new_rj,new_bi,new_bj,cnt+1);
        }
    }

    return -1;
}


int main() {
    freopen("input.txt","r",stdin);

    cin >> N >> M;
    for(int i=0;i<N;i++){
        for(int j=0;j<M;j++){
            cin >> map_arr[i][j];
            if (map_arr[i][j] == 'R') {red_ball[0]=i; red_ball[1]=j; map_arr[i][j] = '.';}
            if (map_arr[i][j] == 'B') {blue_ball[0]=i; blue_ball[1]=j; map_arr[i][j] = '.';}
        }
    }

    printMap(red_ball[0],red_ball[1],blue_ball[0],blue_ball[1]);
    move(red_ball[0],red_ball[1],blue_ball[0],blue_ball[1],1);
    cout << (ans != 11 ? ans : -1);

    return 0;
}


